signac/core/collection_api.py

```diff
@@ -0,0 +1,155 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from contextlib import contextmanager
+from abc import abstractmethod
+from abc import ABCMeta
+from collections import defaultdict
+from collections.abc import Collection
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+
+class SyncedCollectionABCMeta(ABCMeta):
+    """ Metaclass for the definition of SyncedCollection.
+
+    This metaclass automatically registers synced data structures' definition,
+    this is used when recursively converting synced data structures to determine
+    what to convert their children into.
+    """
+    def __init__(cls, name, bases, dct):
+        if not hasattr(cls, 'registry'):
+            cls.registry = defaultdict(list)
+        else:
+            if not cls.__abstractmethods__:
+                cls.registry[cls.backend].append(cls)
+        return super().__init__(name, bases, dct)
+
+
+class SyncedCollection(Collection, metaclass=SyncedCollectionABCMeta):
+    """The base synced collection represents a collection that is synced with a backend.
+
+    The class is intended for use as an ABC. In addition, it declares abstract
+    methods that must be implemented by any subclass.The SyncedCollection is a
+    :class:`~collections.abc.Collection` where all data is stored persistently in
+    the underlying backend.
+    """
+    backend = None
+
+    def __init__(self, parent=None):
+        self._data = None
+        self._parent = parent
+        self._suspend_sync_ = 0
+
+    @classmethod
+    def from_base(cls, data, backend=None, **kwargs):
+        """This method dynamically resolves the type of object to the
+        corresponding synced collection.
```

Comment:
This doc-string formatting is still invalid. @vyasr Is the linter not working?

Reply:
I was having some issues with linter. It's sorted now. I will correct it in next commit.

Reply:
I see some places where the docstrings are still incomplete. The pydocstyle linter settings on the `numpy_docs` branch is ultimately what we want to match, and I would like to move this PR forward and work with automated tooling to incrementally fix these issues at a later time. Specifically, I expect this to happen by merging `numpy_docs` into the main branch in #357 and then pulling the main branch into `feature/synced_collections` (not on @vishav1771's fork).